<a href="https://colab.research.google.com/github/ManoharRavula/Manohar_INFO5731_-Spring2024/blob/main/INFO5731_Assignment_1_Spring_2024.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **INFO5731 Assignment One**

This assignment includes range of programming tasks and data analysis exercises, aiming to enhance your Python skills and proficiency in string manipulation, numerical computations, and data science. Follow the instructions carefully and design well-structured Python programs to address each question.

Late Submission will have a penalty of 10% reduction for each day after the deadline.

# Question 1 (10 points).

The formula that describes the volume V of a hemisphere with radius r is the following:

$ V=\frac{2}{3}\ast\ \pi\ast\ r^3 $

Write a program to calculate the value of V when r is in the range of even numbers from 1-20. Output the result in the following format:

	r	V
	2	…
	4	…
	6	…
	.
	.
	.
	20	…


In [1]:
# Your code here
import numpy as num

def calhemvolume():
    pi = num.pi
    print("r    V")
    for r in range(2, 21, 2):
        # Calculation of the hemisphere
        V = (2/3) * pi * (r**3)
        print(f"{r}    {V:.2f}")
calhemvolume()



r    V
2    16.76
4    134.04
6    452.39
8    1072.33
10    2094.40
12    3619.11
14    5747.02
16    8578.64
18    12214.51
20    16755.16


# Question 2 (20 points)

(a) Write a Python program to generate a sequence of 12 random numbers falling within the range of 50 to 150.

(b) Arrange the generated numbers in ascending order. Display the sorted list.

(c) Extend the program to arrange the numbers in descending order as well. Present the sorted list accordingly.

(d) Compute and output the average of the entire group of generated numbers.

(e) Enhance the program to calculate and display the average of the first 6 numbers from both the ascending and descending sorted lists.

In [1]:
# Write code for each of the sub parts with proper comments.
import random

# (a) Generating the sequence of 12 different numbers within the range of 50 to 150
randomNumbers = [random.randint(50, 150) for _ in range(12)]

# (b) Arranging the numbers in ascending order
ascendingOrder = sorted(randomNumbers)
print(f"Ascending order of numbers: {ascendingOrder}")

# (c) Arranging the numbers in descending order
descendingOrder = sorted(randomNumbers, reverse=True)
print(f"Descending order of numbers: {descendingOrder}")

# (d) compute and average of the numbers
average = sum(randomNumbers) / len(randomNumbers)
print(f"Average of all numbers: {average:.2f}")

# (e) Calculate and display the average of the first 6 numbers from both sorted lists
avgfirstsixascending = sum(ascendingOrder[:6]) / 6
avgfirstsixdescending = sum(descendingOrder[:6]) / 6
print(f"Average of the first 6 numbers (ascending): {avgfirstsixascending:.2f}")
print(f"Average of the first 6 numbers (descending): {avgfirstsixdescending:.2f}")



Ascending order of numbers: [50, 50, 57, 64, 105, 115, 122, 139, 140, 140, 143, 146]
Descending order of numbers: [146, 143, 140, 140, 139, 122, 115, 105, 64, 57, 50, 50]
Average of all numbers: 105.92
Average of the first 6 numbers (ascending): 73.50
Average of the first 6 numbers (descending): 138.33


# Question 3 (20 points).

Create a Python program that encompasses various string manipulation task:

(a) Prompt the user to input two sentences.

(b) Transform each sentence into a list, where each element corresponds to a word in the sentence. For example, if the sentences are "Hello, how are you doing today?" and "This is an example sentence for the program!"

The program should output:
['Hello', ',', 'how', 'are', 'you', 'doing', 'today', '?', 'This', 'is', 'an', 'example', 'sentence', 'for', 'the', 'program', '!']

(c) Compare the two sentences. Print out a list of words that occur in both sentences.

(d) Remove any punctuation from the sentence and print the sanitized sentences. It should remove at least the following characters, but it can remove more: period(“.”), comma (“,”), semicolon (“;”), and colon (“:”).

(e) Count the number of vowels in each of the sentences. Print out each vowel and the occurrences  in the each of the two sentences separately, such as:

Sentence1: a:2, e:2, i:0, etc.
Sentence2: a:1, e:2, i:2, etc.

Design a well-structured Python program to execute these tasks efficiently.

In [3]:
# Your code here
import re
from collections import Counter

# (a) Asking the user to input two sentences.
firstSentence = input("Enter the first sentence: ")
secondSentence = input("Enter the second sentence: ")

# (b) Transforming each sentence to a list of words and punctuation.
def transformSentence(sentence):
    return re.findall(r'\w+|[.,!?;:]', sentence)

firstSentenceWords = transformSentence(firstSentence)
secondSentencewords = transformSentence(secondSentence)
print("Transformed sentences into lists:", firstSentenceWords + secondSentencewords)

# (c) Comparing two sentences and printing out matched words in both sentences
common_words = set(firstSentenceWords).intersection(set(secondSentencewords))
print("Common words in both sentences:", common_words)

# (d) Removing punctuation from the sentences
def removePunctuation(sentence):
    return re.sub(r'[.,!?;:]', '', sentence)

cleanFirstSentence = removePunctuation(firstSentence)
cleanSecondSentence = removePunctuation(secondSentence)
print("cleaned sentence 1:", cleanFirstSentence)
print("cleaned sentence 2:", cleanSecondSentence)

# (e) Counting the number of vowels in each of the sentences.
def count_vowels(sentence):
    vowels = 'aeiouAEIOU'
    counter = Counter([char for char in sentence if char in vowels])
    return dict(counter)

vowel_count1 = count_vowels(cleanFirstSentence)
vowel_count2 = count_vowels(cleanSecondSentence)
print("Vowel occurrences in sentence 1:", vowel_count1)
print("Vowel occurrences in sentence 2:", vowel_count2)



Enter the first sentence: What is your name?
Enter the second sentence: My name is Manohar
Transformed sentences into lists: ['What', 'is', 'your', 'name', '?', 'My', 'name', 'is', 'Manohar']
Common words in both sentences: {'is', 'name'}
cleaned sentence 1: What is your name
cleaned sentence 2: My name is Manohar
Vowel occurrences in sentence 1: {'a': 2, 'i': 1, 'o': 1, 'u': 1, 'e': 1}
Vowel occurrences in sentence 2: {'a': 3, 'e': 1, 'i': 1, 'o': 1}


# Question 4 (20 Points)

Design and implement a simple command-line calculator program. The program should handle both integers and floating-point numbers.

(a) Write program that takes a mathematical expression as input and evaluates it. The program should support addition, subtraction, multiplication, and division.

(b) Extend the calculator to handle parentheses in mathematical expressions. Implement the necessary logic to evaluate expressions with nested parentheses.

(c) Enhance your calculator to include additional mathematical operations such as exponentiation, square root, and modulus.

(d) Implement error handling in the calculator program. Handle cases such as division by zero and invalid input expressions. Provide meaningful error messages.

(e) Implement a command to exit the calculator.

In [5]:
# Your code here
import math

def calculator(expression):
    # Limiting to use the functions and constants from the math module only
    allowedFuntions = {name: obj for name, obj in math.__dict__.items() if not name.startswith("__")}
    # Additional allowed names/functions
    allowedFuntions.update({
        "sqrt": math.sqrt,
        "pow": pow,
        "abs": abs
    })

    try:
        return eval(expression, {"__builtins__": None}, allowedFuntions)
    except ZeroDivisionError:
        return "Error: Division by zero."
    except Exception as e:
        return f"Error: {str(e)}"

def calculate():
    print("Command-Line Calculator")
    print("Supports addition,subtraction,multiplication,division,power-**, modulus-% and sqaute root - sqrt()")
    print("To Exit type exit")

    while True:
        expression = input("Enter numbers to calculate: ").strip()
        if expression.lower() == 'exit':
            print("Exiting calculator.")
            break
        if expression:
            result = calculator(expression)
            print("Result:", result)
# checks if the script is running as the main program
if __name__ == "__main__":
    calculate()


Command-Line Calculator
Supports addition,subtraction,multiplication,division,power-**, modulus-% and sqaute root - sqrt()
To Exit type exit
Enter numbers to calculate: 2+2
Result: 4
Enter numbers to calculate: 2*3
Result: 6
Enter numbers to calculate: sqrt(25)
Result: 5.0
Enter numbers to calculate: 2**3
Result: 8
Enter numbers to calculate: 4/0
Result: Error: Division by zero.
Enter numbers to calculate: 4 divide 4
Result: Error: invalid syntax (<string>, line 1)
Enter numbers to calculate: exit
Exiting calculator.


# Question 5 (30 points).

In the field of Data Science, data is often formatted as a comma-delimited (CSV) file, where each line in a file corresponds to a field's value. Refer to Chapter 9 (p. 106) for more information on comma-delimited files.

To answer this question, you will need to download the file [Assignment1_denton_housing.csv](https://raw.githubusercontent.com/unt-iialab/info5731-spring2022/main/assignments/Assignment1_denton_housing.csv) included with this assignment. The file contains statistics about housing in the city of Denton from 2008 to 2014. Write a program to implement the following questions.

(a) (5 pts) Calculate the difference in the number of Occupied Housing Units from year to year and print it. The difference must be calculated for the consecutive years such as 2008-2009, 2009-2010 etc. Finally, print the values in the ascending order.

(b) (5 pts) For all the years, calculate the percentage of housing units which are vacant an occupied. Print the results in the following format:

Year Vacant Housing Units Occupied Housing Units

    2008  30%   70%                          
	2009 ----- -----
	2010 ----- -----
    2011 ----- -----
	2012 ----- -----
    2013 ----- -----
	2014 ----- -----
  
  

(c) (10 pts) Calculate and print the valued and years in which the highest number of housing units were vacant and occupied. Print the results in the following format:  

                              Year	Value
	Vacant Housing Units	  -----   -----
	Occupied Housing Units	-----   -----

(d) (10 pts) Calculate the harmonic mean of the total housing units and print it out. You can find the information about harmonic mean here: https://ncalculators.com/statistics/harmonic-mean-calculator.htm

In [9]:
 # Your code here
from google.colab import files
uploaded = files.upload()
import csv
from collections import defaultdict

def readHousingDataFile(filename):
    with open(filename, mode='r') as file:
        reader = csv.DictReader(file)
        data = [row for row in reader]
    return data

def calyearlyDifferences(data):
    occupied = {int(row['year']): int(row['value']) for row in data if row['title_field'] == 'Occupied Housing Units'}
    sorted_years = sorted(occupied.keys())
    differences = [(sorted_years[i+1], occupied[sorted_years[i+1]] - occupied[sorted_years[i]]) for i in range(len(sorted_years)-1)]
    return sorted(differences, key=lambda x: x[1])

def calVancyRates(data):
    total = {int(row['year']): int(row['value']) for row in data if row['title_field'] == 'Total Housing Units'}
    occupied = {int(row['year']): int(row['value']) for row in data if row['title_field'] == 'Occupied Housing Units'}
    vacancy_rates = {year: ((total[year] - occupied[year]) / total[year] * 100, occupied[year] / total[year] * 100) for year in total}
    return vacancy_rates

file = 'Assignment1_denton_housing.csv'
data = readHousingDataFile(file)

differences = calyearlyDifferences(data)
print("A) Number of Occupied Housing Units from year to year")
for year, diff in differences:
    print(f"{year-1}-{year}: {diff}")

rates = calVancyRates(data)
print("B) Percentage of housing units which are vacant and occupied")
for year in sorted(rates.keys()):
    vacant, occupied = rates[year]
    print(f"{year} {vacant:.2f}% {occupied:.2f}%")

def findOccuipedunits(data):
    occupiedDict = {int(row['year']): int(row['value']) for row in data if row['title_field'] == 'Occupied Housing Units'}
    vacantDict = {int(row['year']): int(row['value']) for row in data if row['title_field'] == 'Vacant Housing Units'}

    maxOccupiedYear = max(occupiedDict, key=occupiedDict.get)
    maxVacantYear = max(vacantDict, key=vacantDict.get)

    return {
        'Vacant Housing Units': (maxVacantYear, vacantDict[maxVacantYear]),
        'Occupied Housing Units': (maxOccupiedYear, occupiedDict[maxOccupiedYear])
    }

max_units = findOccuipedunits(data)
print(f"c) Value and years in which the highest number of housing units were vacant and occupied")
print("                          Year    Value")
for category, info in max_units.items():
    print(f"{category:25} {info[0]}   {info[1]}")

from statistics import harmonic_mean

def calculate_harmonic_mean(data):
    total_units = [int(row['value']) for row in data if row['title_field'] == 'Total Housing Units']
    return harmonic_mean(total_units)

harmonic_mean_value = calculate_harmonic_mean(data)
print(f"D) Harmonic Mean of total Housing units: {harmonic_mean_value:.2f}")




Saving Assignment1_denton_housing.csv to Assignment1_denton_housing (3).csv
A) Number of Occupied Housing Units from year to year
2008-2009: -795
2010-2011: -303
2011-2012: 1404
2012-2013: 1565
2013-2014: 2622
2009-2010: 5091
B) Percentage of housing units which are vacant and occupied
2008 10.88% 89.12%
2009 9.07% 90.93%
2010 8.14% 91.86%
2011 9.38% 90.62%
2012 6.68% 93.32%
2013 8.80% 91.20%
2014 5.73% 94.27%
c) Value and years in which the highest number of housing units were vacant and occupied
                          Year    Value
Vacant Housing Units      2008   4483
Occupied Housing Units    2014   46295
D) Harmonic Mean of total Housing units: 44398.29


# Mandatory Question

Provide your thoughts on the assignment. What did you find challenging, and what aspects did you enjoy? Your opinion on the provided time to complete the assignment. Share any suggestions for improvement. Your feedback will be valuable for enhancing the learning experience.

In [ ]:
# Write your response below
#I have found this assignment tough and challenging,  I really liked learning how to import and handle data in Python, especially doing things like reading from a CSV file and processing the data and analyzing it. Also the calculator problem was something that i though python couldnt do it but using the right modules and built in functions it went good.Doing these kinds of assignments is really helping to broaden my programming skills and what all i can do.